HSS/LMS Crypto Documentation (#140)
sehlen-bsi/botan-docs: docs/cryptodoc/src/05_06_hss_lms.rst

.. _pubkey/hss_lms:

HSS/LMS
=======

Botan implements the Hierarchical Signature System (HSS) with Leighton-Micali
Hash-Based Signatures (HSS/LMS) as defined in [RFC8554]_. It supports the
parameter sets defined in [RFC8554]_ and those in [draft-fluhrer-11]_.

Algorithm Internals
-------------------

HSS/LMS consists of three building blocks. Like most hash-based signature schemes,
it uses a One-Time Signature (OTS) at its base, named Leighton-Micali OTS
(LM-OTS). The public keys of multiple LM-OTS instances are processed to the leaves

> of a Merkle tree. This composition is called the Leighton-Micali Signature (LMS)

Author: I didn't want to say "... are the public keys" because they were hashed before. But I agree that "are processed" sounds somehow weird. What do you think of `The public keys of multiple LM-OTS instances compose the leaves of a Merkle tree.`?

```

> of a Merkle tree. This composition is called the Leighton-Micali Signature (LMS)

Reviewer: Sounds better to me

   +---------------------------------------+--------------------------------------------+---------------------------+-----------------------+
   | :ref:`LMS <pubkey/hss_lms/lms>`       | :srcref:`src/lib/pubkey/hss_lms/lms.h`     | LMS                       | 5.                    |
   +---------------------------------------+--------------------------------------------+---------------------------+-----------------------+
   | :ref:`HSS <pubkey/hss_lms/hss>`       | :srcref:`src/lib/pubkey/hss_lms/hss.h`     | HSS with LMS              | 6.                    |
   +---------------------------------------+--------------------------------------------+---------------------------+-----------------------+
   | HSS/LMS                               | :srcref:`src/lib/pubkey/hss_lms/hss_lms.h` | Botan's HSS/LMS interface |                       |
   +---------------------------------------+--------------------------------------------+---------------------------+-----------------------+

.. _pubkey/hss_lms/lm_ots:

LM-OTS
^^^^^^

LM-OTS is configured with several parameters. The first parameter is the used hash
function. Botan's implementation only allows one hash function for all

> LMS trees and their LM-OTS algorithm (recommended in [SP800-208]_). The width of

Author: Yeah, you are right. I thought RFC 8554 did not mention it because it only defines a single hash function. But they do!

the Winternitz coefficient ``w`` is the second parameter, defining the
time-signature-size-tradeoff of the LM-OTS instance.
Those first two parameters implicitly define the hash function output size ``n``,
the number of Winternitz chains ``p``, and the constant ``ls`` used for the
checksum computation (see [RFC8554]_ Section 4.1.). Botan allows all combinations
of hash function and ``w`` defined in [RFC8554]_ and [draft-fluhrer-11]_, which
are listed in Table :ref:`Supported LM-OTS parameter sets <pubkey/hss_lms/lm-ots-params>`.

.. _pubkey/hss_lms/lm-ots-params:

> .. table::  Supported LM-OTS parameter sets

Author: Yeah. However, I think we should either remove all tables or none.

> .. table::  Supported LM-OTS parameter sets

Author: Again, I think all or nothing.

> .. table::  Supported LM-OTS parameter sets

Reviewer: I agree, was more a reminder in case a decision is made before this gets merged.

> .. table::  Supported LM-OTS parameter sets

Reviewer: I think this is a duplicate of my other comments thanks to the "nice" internet in the trains 😅.

   +---------------------+--------------+-----+-----+-----+-----+--------+
   | LMOTS_SHAKE_N32_W8  | SHAKE256/256 | 32  | 8   | 34  | 0   | 0x000c |
   +---------------------+--------------+-----+-----+-----+-----+--------+
   | LMOTS_SHAKE_N24_W1  | SHAKE256/192 | 24  | 1   | 200 | 8   | 0x000d |
   +---------------------+--------------+-----+-----+-----+-----+--------+
   | LMOTS_SHAKE_N24_W2  | SHAKE256/192 | 24  | 2   | 101 | 6   | 0x000e |
   +---------------------+--------------+-----+-----+-----+-----+--------+
   | LMOTS_SHAKE_N24_W4  | SHAKE256/192 | 24  | 4   | 51  | 4   | 0x000f |
   +---------------------+--------------+-----+-----+-----+-----+--------+
   | LMOTS_SHAKE_N24_W8  | SHAKE256/192 | 24  | 8   | 26  | 0   | 0x0010 |
   +---------------------+--------------+-----+-----+-----+-----+--------+

In addition to these parameters, an LM-OTS
instance is defined by the identifier ``I`` of the LMS tree
and the index of its leaf ``q``, where the LM-OTS instance is located; this is
represented by the class ``OTS_Instance``.
For each LM-OTS instance, we can create a keypair with a secret key (class
``LMOTS_Private_Key``) and a public key (class ``LMOTS_Public_Key``). The

> pseudorandom key generation method recommended in [RFC8554]_ Appendix A derives the

Author: Ok, I'll add a reference to SP800-208. However, since this document is called _Recommendation for Stateful Hash-Based Signature Schemes_, the term "recommended" should be fine.

> pseudorandom key generation method recommended in [RFC8554]_ Appendix A derives the

Reviewer: I looked it up again to be sure, and only found a *shall* in *6.1 LMS and HSS Random Number Generation Requirements* of the NIST document for this.

So I was wrong, Probably still good to keep the reference to SP800-208, but make it recommended only.

secret key's Winternitz chain inputs (``x[]`` of [RFC8554]_). The inputs for this
method are the LM-OTS instance parameters and a
secret seed ``SEED`` associated with an LMS tree:

.. math::
   \mathtt{x[i]\ =\ Hash(I\ ||\ u32str(q)\ ||\ u16str(i)\ ||\ u8str(0xff)\ ||\ SEED)}

The public key is created by computing all Winternitz hash chains
beginning with their secret chain inputs ``x[]`` (see [RFC8554]_ Algorithm 1).
Besides the instance parameters, it contains the final hash value denoted as
``K`` in [RFC8554]_ Algorithm 1.

For creating an LM-OTS signature of a message, Botan offers the method
``LMOTS_Private_Key::sign``. For that, it implements Algorithm 1 of [RFC8554]_.
One important remark is the creation of the randomizer ``C``. To create this

> randomizer, Botan adapts the same approach as the reference implementation by

Author: Well, the implementation is mentioned in [RFC 8554 Appendix E](https://datatracker.ietf.org/doc/html/rfc8554#appendix-E), so I think it can be called somehow official. I can mention Cisco, though.

> randomizer, Botan adapts the same approach as the reference implementation by

Reviewer: Missed that it is mentioned in the Appendix, I thought it was only referenced for the measurements taken.

HSS/LMS secret key format is not defined in [RFC8554]_. Botan defines the following
secret key byte format written in the same syntax as [RFC8554]_:

.. math::
   \mathtt{SK\_Bytes =\ } &\mathtt{u32str(L)\ ||\ u64str(idx)\ || }

      &\mathtt{u32str(LMSAlgorithmId_{root\_layer})\ ||\ u32str(LMOTSAlgorithmId_{root\_layer})\ || }

      &\mathtt{\dots\ || }

      &\mathtt{u32str(LMSAlgorithmId_{bottom\_layer})\ ||\ u32str(LMOTSAlgorithmId_{bottom\_layer})\ || }

      &\mathtt{SEED_{root\_tree} ||\ I_{root\_tree} }

``idx`` is the index of the next signature created using this
secret key, defining the LMS leaves to use. This entry updates
after every signature creation. Next, the LMS and LMOTS algorithm IDs are given
for each level as defined in Tables :ref:`Supported LM-OTS parameter sets
<pubkey/hss_lms/lm-ots-params>` and :ref:`Supported LMS parameter sets
<pubkey/hss_lms/lms-params>`. Finally, ``SEED`` and ``I`` of the root LMS tree
are given. The classes ``HSS_LMS_PublicKeyInternal`` and
``HSS_LMS_PrivateKeyInternal`` realize the public and secret key, respectively.

Botan's HSS/LMS implementation derives LMS seeds and identifiers

> by the same method the reference implementation applies.

Author: Yeah, why not.

.. _pubkey/hss_lms:

HSS/LMS
=======

Botan implements the Hierarchical Signature System (HSS) with Leighton-Micali
Hash-Based Signatures (HSS/LMS) as defined in [RFC8554]_ under consideration of
the recommendations of [SP800-208]_. It supports the parameter sets defined in
[RFC8554]_ and those in [draft-fluhrer-11]_.

Algorithm Internals
-------------------

The Hierarchical Signature System (HSS) with Leighton-Micali

> Hash-Based Signatures (HSS/LMS) consists of three building blocks.

Author: Meant are OTS, LMS, and HSS. The term _building blocks_ may be misleading. I'll go with _multiple components_.

Like most hash-based signature schemes,
it uses a One-Time Signature (OTS) at its base, named Leighton-Micali OTS
(LM-OTS). The public keys of multiple LM-OTS instances compose the leaves
of a Merkle tree. This composition is the basis of the Leighton-Micali Signature (LMS)
method. The root node of the LMS Merkle tree defines its public key. [RFC8554]_
also provides HSS, a hypertree composition of multiple LMS trees, where the leaves

> of LMS trees sign the public keys of other LMS trees.

Reviewer 2: ```suggestion
of LMS trees sign the public keys of subsequent LMS trees.
```

   +---------------------------------------+--------------------------------------------+---------------------------+-----------------------+
   | HSS/LMS                               | :srcref:`src/lib/pubkey/hss_lms/hss_lms.h` | Botan's HSS/LMS interface |                       |
   +---------------------------------------+--------------------------------------------+---------------------------+-----------------------+

.. _pubkey/hss_lms/lm_ots:

LM-OTS
^^^^^^

LM-OTS is configured with several parameters. The first parameter is the used hash
function. Botan's implementation only allows one hash function for all
LMS trees and their LM-OTS algorithm (recommended in [RFC8554]_ and [SP800-208]_).
The width of the Winternitz coefficient ``w`` is the second parameter, defining the
time-signature-size-tradeoff of the LM-OTS instance.
Those first two parameters implicitly define the hash function output size ``n``,
the number of Winternitz chains ``p``, and the constant ``ls`` used for the
checksum computation (see [RFC8554]_ Section 4.1.). Botan allows all combinations

> of hash function and ``w`` defined in [RFC8554]_ and [draft-fluhrer-11]_, which
> are listed in Table :ref:`Supported LM-OTS parameter sets <pubkey/hss_lms/lm-ots-params>`.

Reviewer 2: ```suggestion
of hash function and ``w`` that are defined in [RFC8554]_ and [draft-fluhrer-11]_ and listed here
in Table :ref:`Supported LM-OTS parameter sets <pubkey/hss_lms/lm-ots-params>`.
```
an editorial suggestion

Author: That sounds better, thanks!

and the index of its leaf ``q``, where the LM-OTS instance is located; this is
represented by the class ``OTS_Instance``.
For each LM-OTS instance, we can create a keypair with a secret key (class
``LMOTS_Private_Key``) and a public key (class ``LMOTS_Public_Key``). As recommended
by [SP800-208]_, Botan uses the pseudorandom key generation method of [RFC8554]_ Appendix A to
derive the secret key's Winternitz chain inputs (``x[]`` of [RFC8554]_). The inputs for this
method are the LM-OTS instance parameters and a
secret seed ``SEED`` associated with an LMS tree:

.. math::
   \mathtt{x[i]\ =\ Hash(I\ ||\ u32str(q)\ ||\ u16str(i)\ ||\ u8str(0xff)\ ||\ SEED)}

The public key is created by computing all Winternitz hash chains
beginning with their secret chain inputs ``x[]`` (see [RFC8554]_ Algorithm 1).
Besides the instance parameters, it contains the final hash value denoted as
``K`` in [RFC8554]_ Algorithm 1.

For creating an LM-OTS signature of a message, Botan offers the method
``LMOTS_Private_Key::sign``. For that, it implements Algorithm 1 of [RFC8554]_.
One important remark is the creation of the randomizer ``C``. To create this

> randomizer, Botan adapts the same approach as the Cisco reference implementation

Reviewer 2: Maybe add a reference to the github repo of the Cisco reference implementation?

Author: Sure. I'll reference [RFC 8554 Appendix E](https://datatracker.ietf.org/doc/html/rfc8554#appendix-E), which contains a link to the implementation and calls it "an example implementation"

``LMS_Private_Key::sign_and_pk_gen``, which signs the message and computes the
public key associated with the LMS instance according to Section 5.3. and 5.4.
of [RFC8554]_. For verification of a signature-message pair, Botan provides
``LMS_PublicKey::verify_signature``, implementing  Algorithm 5 of [RFC8554]_.
The internal logic to create and reconstruct Merkle trees is implemented in the
cross-algorithm helper module ``tree_hash``
(:srcref:`src/lib/utils/tree_hash/tree_hash.h`). The tree's leaves are created using the
constructs introduced in :ref:`Section LM-OTS <pubkey/hss_lms/lm_ots>`.

.. _pubkey/hss_lms/hss:

HSS
^^^

Multiple LMS trees build up an HSS/LMS hypertree, where leaf nodes of higher LMS
trees sign the public keys of lower LMS instances. The following
parameters define the HSS/LMS hypertree. The parameter ``L`` configures the height
of the HSS/LMS hypertree, i.e., the number of LMS tree levels in the hypertree.
As specified in [RFC8554]_, Botan allows ``L`` to be within one and eight.
An LMS and LM-OTS parameter set pair is defined for each level. Botan allows all
parameter combinations as long as the hash function

> is always the same in every set at every level.

Reviewer 2: Why? Does this follow the philosophy of the RFC? If yes, then better state it here, so that it does not appear as a random choice.

Author: Yeah, it is a NIST SP 800-208 recommendation:
>If the HSS instance has more than one level, then the hash function used for the tree at level 0 shall be used for every LMS tree at every other level.

I agree, a reference is missing here!

pseudorandom key generation method of [RFC8554]_ Appendix A.
The derivation functions are the following:

.. math::
   \mathtt{SEED_{child}}\ &\mathtt{= Hash(I_{parent}\ ||\ u32str(q_{parent})\
   ||\ u16str(0xfffe)\ ||\ u8str(0xff)\ ||\ SEED_{parent})}

   \mathtt{I_{child}}\    &\mathtt{= Hash(I_{parent}\ ||\ u32str(q_{parent})\
   ||\ u16str(0xffff)\ ||\ u8str(0xff)\ ||\ SEED_{parent})}

:math:`\mathtt{I_{parent}}` and :math:`\mathtt{SEED_{parent}}` are the
identifier and seed of the parent LMS tree, while :math:`\mathtt{I_{child}}`
and :math:`\mathtt{SEED_{child}}` are derived for the new child.
:math:`\mathtt{q_{parent}}` is the parent's LMS leaf index used to sign the
child LMS tree's public key. Note that since the third entry separates their
domain, the hash inputs will never collide with
the ones of the LM-OTS computations of ``x[i]`` and ``C``.

The method ``HSS_LMS_Signature_Operation::sign`` is used for signature creation,
implementing Algorithm 8 of [RFC8554]_.
``HSS_LMS_Verification_Operation::is_valid_signature`` provides signature
verification, as in Section 6.3. of [RFC8554]_.

.. _pubkey/hss_lms/key_gen:

Key Generation
--------------

> HSS/LMS key generation follows Section 6.1. of [RFC8554]_ and is implemented

Reviewer 2: I think, if you state here that Botan follows §6.1 in RFC8554 but above in line 263 ff that "Botan's HSS/LMS implementation derives LMS seeds and identifiers by the same method Cisco's reference implementation applies.", then you should comment on the compatibility with the sentence

"In the above algorithm, each LMS public/private key pair generated MUST be generated independently."

of §6.1 in RFC8554. Otherwise people might get confused or ask.

within the ``HSS_LMS_PrivateKeyInternal`` constructor and
``HSS_LMS_PublicKeyInternal::create``. It works as follows:

.. admonition:: HSS/LMS Key Generation

   **Input:**

   -  ``rng``: random number generator
   -  ``L``: The number of levels in the HSS/LMS hypertree
   -  ``lms-params[0], ..., lms-params[L-1]`` : LMS parameter sets at all
      ``L`` levels
   -  ``lm-ots-params[0], ..., lm-ots-params[L-1]``: LM-OTS parameter sets at all
      ``L`` levels

   **Output:**

   -  ``SK``, ``PK``: secret and public key

   **Steps:**

   1. Generate new values ``SEED`` and ``I`` using ``rng``.
   2. ``idx = 0`` sets the initial signature index.
   3. | ``SK = {L, idx, lms-params[0], lm-ots-params[0], ..., lms-params[L-1],``
      |       ``lm-ots-params[L-1], SEED, I}``.
   4. Construct the root LMS secret key ``lms-sk[0]`` with parameters
      ``lms-params[0]`` and ``lm-ots-params[0]`` containing ``SEED`` and ``I``
      (see :ref:`LMS <pubkey/hss_lms/lms>`).
   5. Construct the root LMS public key ``lms-pk[0]`` from ``lms-sk[0]``
      (see :ref:`LMS <pubkey/hss_lms/lms>`).
   6. ``PK = {L, lms-pk[0]}``.

>    **Notes:**

Reviewer 2: Should the notes also reflect that the given parameters for lm-ots and lms for the different levels are checked concerning the restrictions imposed? i.e. matching hash function etc.

   ||\ u16str(0xfffe)\ ||\ u8str(0xff)\ ||\ SEED_{parent})}

   \mathtt{I_{child}}\    &\mathtt{= Hash(I_{parent}\ ||\ u32str(q_{parent})\
   ||\ u16str(0xffff)\ ||\ u8str(0xff)\ ||\ SEED_{parent})}

:math:`\mathtt{I_{parent}}` and :math:`\mathtt{SEED_{parent}}` are the
identifier and seed of the parent LMS tree, while :math:`\mathtt{I_{child}}`
and :math:`\mathtt{SEED_{child}}` are derived for the new child.
:math:`\mathtt{q_{parent}}` is the parent's LMS leaf index used to sign the
child LMS tree's public key. Note that since the third entry separates their
domain, the hash inputs will never collide with
the ones of the LM-OTS computations of ``x[i]`` and ``C``.

The method ``HSS_LMS_Signature_Operation::sign`` is used for signature creation,
implementing Algorithm 8 of [RFC8554]_.
``HSS_LMS_Verification_Operation::is_valid_signature`` provides signature
verification, as in Section 6.3. of [RFC8554]_.

.. _pubkey/hss_lms/key_gen:

Key Generation
--------------

HSS/LMS key generation follows Section 6.1. of [RFC8554]_ and is implemented
within the ``HSS_LMS_PrivateKeyInternal`` constructor and
``HSS_LMS_PublicKeyInternal::create``.

> Note that [RFC8554]_ and [SP800-208]_ require that each LMS public/private key

Reviewer 3: ```suggestion
Note that [RFC8554]_ and [SP800-208]_ require that all LMS instances' public/private key pairs
```

Botan's key generation algorithm works as follows:

.. admonition:: HSS/LMS Key Generation

   **Input:**

   -  ``rng``: random number generator
   -  ``L``: The number of levels in the HSS/LMS hypertree
   -  ``lms-params[0], ..., lms-params[L-1]`` : LMS parameter sets at all
      ``L`` levels
   -  ``lm-ots-params[0], ..., lm-ots-params[L-1]``: LM-OTS parameter sets at all
      ``L`` levels

   **Output:**

   -  ``SK``, ``PK``: secret and public key

   **Steps:**

   1. Generate new values ``SEED`` and ``I`` using ``rng``.
   2. ``idx = 0`` sets the initial signature index.
   3. | ``SK = {L, idx, lms-params[0], lm-ots-params[0], ..., lms-params[L-1],``
      |       ``lm-ots-params[L-1], SEED, I}``.
   4. Construct the root LMS secret key ``lms-sk[0]`` with parameters
      ``lms-params[0]`` and ``lm-ots-params[0]`` containing ``SEED`` and ``I``
      (see :ref:`LMS <pubkey/hss_lms/lms>`).
   5. Construct the root LMS public key ``lms-pk[0]`` from ``lms-sk[0]``
      (see :ref:`LMS <pubkey/hss_lms/lms>`).
   6. ``PK = {L, lms-pk[0]}``.

   **Notes:**

   - A formatted string provides ``L`` and the LMS and OTS parameters.

>    - In contrast to [RFC8554]_ Algorithm 7. Step 2, the keys and signatures of

Reviewer 3: ```suggestion
   - In contrast to [RFC8554]_ Algorithm 7 Step 2, the keys and signatures of
```

   **Input:**

   -  ``m``: message to be signed
   - | ``SK``: HSS/LMS secret key, ``SK = {L, idx, lms-params[0], lm-ots-params[0], ...,``
     |      ``lms-params[L-1], lm-ots-params[L-1], SEED, I}``

   **Output:**

   -  ``sig``:  HSS/LMS signature

   **Steps:**

   1. If ``idx`` denotes that ``SK`` is exhausted, the signature creation is
      aborted.
   2. Derive the LMS signing leaf indices ``q[0], ..., q[L-1]`` from ``idx`` and
      the LMS parameters.
   3. Derive the LMS secret keys ``lms-sk[i]`` for HSS/LMS levels
      ``i = 1, ..., (L-1)`` using the seed and identifier derivation method
      described in :ref:`HSS <pubkey/hss_lms/hss>`.
   4. ``lms-sig[L-1], lms-pk[L-1] = lms-sk[L-1].sign_and_pk_gen(msg, q[L-1])``
      creates the bottom layer LMS signature and the public key bytes of the
      signing LMS tree.
   5. ``lms-sig[i], lms-pk[i] = lms-sk[i].sign_and_pk_gen(lms-pk[i+1], q[i])``
      creates the higher level public key signatures and public keys for
      ``i = L-2, ..., 0``.
   6. | ``sig = u32str(L-1) || lms-sig[0] || lms-pk[1] || lms-sig[1] || ...``
      |       ``|| lms-pk[L-1] || lms-sig[L-1]``.

   **Notes:**

>    - After signature creation, ``idx`` of ``SK`` increased by one.

Reviewer 3: ```suggestion
   - After signature creation, ``idx`` of ``SK`` is increased by one.
```

.. _pubkey/hss_lms/sig_validation:

Signature Validation
--------------------

Botan's method ``HSS_LMS_Verification_Operation::is_valid_signature`` verifies a
signature-message pair by implementing the method of Section 6.3. of [RFC8554]_.
It does the following:

.. admonition:: HSS/LMS Signature Validation

   **Input:**

   -  ``m``: message to be validated
   -  ``sig``: signature to be validated
   -  ``PK``: HSS/LMS public key, ``PK = {L, lms-pk[0]}``

   **Output:**

>    -  ``true`` if the signature for message ``m`` is valid. ``false`` otherwise

Reviewer 3: ```suggestion
   -  ``true`` if the signature for message ``m`` is valid. ``false`` otherwise.
```

defined by its identifier ``I`` and the LM-OTS parameters used for all
contained LM-OTS instances.
We can create a keypair with a secret key (class ``LMS_Private_Key``) and a
public key (class ``LMS_Public_Key``) for each LMS instance. The secret
key contains the value ``SEED`` used for LM-OTS secret key derivation, while the
public key contains the LMS tree's root node. The public key is derived from the
secret key in the constructor of ``LMS_Public_Key``.

For creating an LMS signature, Botan offers the method
``LMS_Private_Key::sign_and_pk_gen``, which signs the message and computes the
public key associated with the LMS instance according to Section 5.3. and 5.4.
of [RFC8554]_. For verification of a signature-message pair, Botan provides
``LMS_PublicKey::verify_signature``, implementing  Algorithm 5 of [RFC8554]_.
The internal logic to create and reconstruct Merkle trees is implemented in the
cross-algorithm helper module ``tree_hash``
(:srcref:`[src/lib]/utils/tree_hash/tree_hash.h`). The tree's leaves are created using the
constructs introduced in :ref:`Section LM-OTS <pubkey/hss_lms/lm_ots>`.

.. _pubkey/hss_lms/hss:

HSS
^^^

> Multiple LMS trees build up an HSS/LMS hypertree, where leaf nodes of higher LMS

Reviewer 4: ```suggestion
An HSS/LMS hypertree consists of multiple LMS trees , where leaf nodes of higher LMS
```

trees sign the public keys of lower LMS instances. The following

> parameters define the HSS/LMS hypertree. The parameter ``L`` configures the height

Reviewer 4: Shouldn't it then always be an HSS hypertree, not HSS/LMS?

of the HSS/LMS hypertree, i.e., the number of LMS tree levels in the hypertree.

> As specified in [RFC8554]_, Botan allows ``L`` to be within one and eight.

Reviewer 4: ```suggestion
As specified in [RFC8554]_, the permitted valuesfor ``L`` in Botan are 1,2,...,8.
```
Maybe you have a better idea.

An LMS and LM-OTS parameter set pair is defined for each level. Botan allows all
parameter combinations as long as the hash function is always the

> same in every set at every level (recommended in [SP800-208]_).

Reviewer 4: ```suggestion
same at all levels (recommended in [SP800-208]_).
```
